File: src/update.ts
/** Update check: prefers Rust backend (reliable network), falls back to GitHub API fetch. */

import { invoke } from "@tauri-apps/api/core";

export const RELEASES_PAGE = "https://github.com/xl495/package-run/releases";
export const LATEST_API =
  "https://api.github.com/repos/xl495/package-run/releases/latest";

const DISMISS_KEY = "package-run-dismissed-update";

export interface AvailableUpdate {
  version: string;
  notes: string;
  htmlUrl: string;
}

/** Compare SemVer (leading `v` optional). Returns true if remote > local. */
export function isNewerVersion(remote: string, local: string): boolean {
  const r = parseSemver(remote);
  const l = parseSemver(local);
  if (!r || !l) return remote.replace(/^v/i, "") !== local.replace(/^v/i, "");
  for (let i = 0; i < 3; i++) {
    if (r[i] > l[i]) return true;
    if (r[i] < l[i]) return false;
  }
  return false;
}

function parseSemver(v: string): [number, number, number] | null {
  const m = v.trim().replace(/^v/i, "").match(/^(\d+)\.(\d+)\.(\d+)/);
  if (!m) return null;
  return [Number(m[1]), Number(m[2]), Number(m[3])];
}

export function getDismissedVersion(): string | null {
  return localStorage.getItem(DISMISS_KEY);
}

export function dismissUpdate(version: string) {
  localStorage.setItem(DISMISS_KEY, version.replace(/^v/i, ""));
}

async function fetchLatestViaJs(): Promise<AvailableUpdate | null> {
  const res = await fetch(LATEST_API, {
    headers: {
      Accept: "application/vnd.github+json",
      "User-Agent": "Package-Run-Updater",
    },
  });
  if (!res.ok) throw new Error(`GitHub API ${res.status}`);
  const data = (await res.json()) as {
    tag_name?: string;
    name?: string;
    body?: string;
    html_url?: string;
    draft?: boolean;
    prerelease?: boolean;
  };
  if (data.draft || data.prerelease) return null;
  const version = (data.tag_name ?? data.name ?? "").replace(/^v/i, "");
  if (!version) return null;
  return {
    version,
    notes: (data.body ?? "").trim(),
    htmlUrl: data.html_url ?? `${RELEASES_PAGE}/tag/v${version}`,
  };
}

async function fetchLatestViaRust(
  currentVersion: string,
): Promise<AvailableUpdate | null> {
  return invoke<AvailableUpdate | null>("check_app_update", {
    currentVersion,
  });
}

export async function checkForUpdate(
  currentVersion: string,
  opts?: { ignoreDismissed?: boolean },
): Promise<AvailableUpdate | null> {
  let latest: AvailableUpdate | null = null;
  try {
    latest = await fetchLatestViaRust(currentVersion);
  } catch {
    const remote = await fetchLatestViaJs();
    if (remote && isNewerVersion(remote.version, currentVersion)) {
      latest = remote;
    }
  }

  if (!latest) return null;
  if (!opts?.ignoreDismissed) {
    const dismissed = getDismissedVersion();
    if (dismissed && dismissed === latest.version.replace(/^v/i, "")) {
      return null;
    }
  }
  return latest;
}

File: src-tauri/src/update.rs
use serde::Serialize;

const LATEST_API: &str = "https://api.github.com/repos/xl495/package-run/releases/latest";
const RELEASES_PAGE: &str = "https://github.com/xl495/package-run/releases";

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AvailableUpdate {
    pub version: String,
    pub notes: String,
    pub html_url: String,
}

#[derive(Debug, serde::Deserialize)]
struct GhRelease {
    tag_name: Option<String>,
    name: Option<String>,
    body: Option<String>,
    html_url: Option<String>,
    draft: Option<bool>,
    prerelease: Option<bool>,
}

fn parse_semver(v: &str) -> Option<(u64, u64, u64)> {
    let s = v.trim().trim_start_matches('v').trim_start_matches('V');
    let mut parts = s.split('.');
    let major = parts.next()?.parse().ok()?;
    let minor = parts.next()?.parse().ok()?;
    let patch = parts
        .next()
        .and_then(|p| {
            let digits: String = p.chars().take_while(|c| c.is_ascii_digit()).collect();
            digits.parse().ok()
        })
        .unwrap_or(0);
    Some((major, minor, patch))
}

/// Returns true if remote is strictly newer than local.
pub fn is_newer(remote: &str, local: &str) -> bool {
    match (parse_semver(remote), parse_semver(local)) {
        (Some(r), Some(l)) => r > l,
        _ => {
            remote.trim().trim_start_matches(['v', 'V'])
                != local.trim().trim_start_matches(['v', 'V'])
        }
    }
}

/// Check GitHub Releases for a newer version than `current_version`.
/// Returns `Ok(None)` when already up to date (or only draft/prerelease).
#[tauri::command]
pub fn check_app_update(current_version: String) -> Result<Option<AvailableUpdate>, String> {
    let body = ureq::get(LATEST_API)
        .set("Accept", "application/vnd.github+json")
        .set("User-Agent", "Package-Run-Updater")
        .timeout(std::time::Duration::from_secs(12))
        .call()
        .map_err(|e| format!("update check failed: {e}"))?
        .into_string()
        .map_err(|e| format!("update response read failed: {e}"))?;

    let release: GhRelease =
        serde_json::from_str(&body).map_err(|e| format!("update parse failed: {e}"))?;

    if release.draft.unwrap_or(false) || release.prerelease.unwrap_or(false) {
        return Ok(None);
    }

    let version = release
        .tag_name
        .or(release.name)
        .unwrap_or_default()
        .trim()
        .trim_start_matches(['v', 'V'])
        .to_string();

    if version.is_empty() {
        return Ok(None);
    }

    if !is_newer(&version, &current_version) {
        return Ok(None);
    }

    let html_url = release
        .html_url
        .unwrap_or_else(|| format!("{RELEASES_PAGE}/tag/v{version}"));

    Ok(Some(AvailableUpdate {
        version,
        notes: release.body.unwrap_or_default().trim().to_string(),
        html_url,
    }))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn semver_newer() {
        assert!(is_newer("0.2.0", "0.1.0"));
        assert!(is_newer("v1.0.0", "0.9.9"));
        assert!(!is_newer("0.1.0", "0.1.0"));
        assert!(!is_newer("0.1.0", "0.2.0"));
    }
}
